infrastructure/src/main/java/org/corfudb/infrastructure/log/CompactionPolicy.java

```diff
@@ -0,0 +1,38 @@
+package org.corfudb.infrastructure.log;
+
+import org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuError;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.util.List;
+
+/**
+ * The interface of stream log compaction policy, which decides
+ * which log segments will be selected for compaction.
+ *
+ * Created by WenbinZhu on 5/22/19.
+ */
+public interface CompactionPolicy {
+
+    /**
+     * Returns a list of segments selected for compaction based on
+     * the compaction policy implementation.
+     *
+     * @param compactibleSegments all unprotected segments that can be selected for compaction.
+     * @return a list of segments selected for compaction.
+     */
+    List<StreamLogSegment> segmentsToCompact(List<StreamLogSegment> compactibleSegments);
+
+    static CompactionPolicy getPolicy(StreamLogParams params) {
```

Comment:
Why use reflection? StreamLogParams could specify the policy in other ways, e.g. Enum.

Reply:
First, the policy will be later specified by user on server startup, it's easier to use string. 
Second, even the StreamLogParams use Enum for policy, the implementation of the policy is not, so we have only two ways: `if...else` or reflection.